### ISSUE TITLE
Nested data source doesn't resolve dataloader in some cases

### ISSUE THREAD

**Issue by mtsmfm** [rmosolgo/graphql-ruby#3706]
**Versions**

`graphql` version: 1.12.19

**Steps to reproduce**
```ruby
require 'bundler/inline'

gemfile do
  source 'https://rubygems.org'

  gem 'graphql', '1.12.19'
end

class Echo < GraphQL::Dataloader::Source
  def fetch(keys)
    keys
  end
end

class Nested < GraphQL::Dataloader::Source
  def fetch(keys)
    dataloader.with(Echo).load_all(keys)
  end
end

class Nested2 < GraphQL::Dataloader::Source
  def fetch(keys)
    dataloader.with(Nested).load_all(keys)
  end
end

class QueryType < GraphQL::Schema::Object
  field :nested, String, null: true
  field :nested2, String, null: true

  def nested
    dataloader.with(Nested).load("nested")
  end
  
  def nested2
    dataloader.with(Nested2).load("nested2")
  end
end

class TestSchema < GraphQL::Schema
  query QueryType
  use GraphQL::Dataloader
end

pp TestSchema.execute(<<~GRAPHQL).to_h
{
  nested
}
GRAPHQL

#=> {"data"=>{"nested"=>"nested"}}

pp TestSchema.execute(<<~GRAPHQL).to_h
{
  nested2
}
GRAPHQL

#=> {"data"=>{"nested2"=>nil}}

pp TestSchema.execute(<<~GRAPHQL).to_h
{
  nested, nested2
}
GRAPHQL

#=> {"data"=>{"nested"=>nil, "nested2"=>nil}}

```

**Expected behavior**

- Field `nested` always returns `"nested"`
- Field `nested2` always returns `"nested2"`

**Actual behavior**

- Field `nested` returns `nil` if it's queried with field `nested2`
- Field `nested2` always returns `nil`

**Comment by rmosolgo** [comment]
Awesome, thanks so much for the detailed report! I'll take a look and follow up here soon.

**Comment by rmosolgo** [comment]
I've got a fix in the works at #3707
